[FR] Improve DAC custom folder init

Mikaayenson opened the pull request:
<!--
Thank you for your interest in and contributing to Detection Rules!
There are a few simple things to check before submitting your pull request
that can help with the review process. You should delete these items
from your submission, but they are here to help bring them to your attention.
-->

## Issues
<!-- Link to related issues. Use closing keywords when appropriate -->

Resolves #3621 

## Summary

Improves the overall experience for users trying to initialize a custom folder using the `python -m detection_rules custom-rules init-config` command. 
- `init-config` --> `setup-config`: Not just base file structure, but setup with initial file content so users can easily get to testing, and removed the config details since we're not defaulting to initialize files with content. Added a cli param to allow users to specify kibana version during setup. Also echoed doc links post setup.
- `SystemExit`: Decided to keep this raised exit since the `RULES_CONFIG = parse_rules_config()` is initialized in so many places in the repo. Also because we dont have CLI context to try/catch.
- Moved `_config.yml` validation to a dataclass (note: couldn't use mixins because of the multiple circular dependancies)
- Added `packages.yaml` validation to the post init of the TestConfig
- Updated doc references as needed.


A few bug fixes:
- The building block folder was missing from the original custom init
- Corrected the example configs to match the paths generated by the init

## Testing

<details><summary>Generated _config.yaml files</summary>
<p>

With defaults:

```
files:
  deprecated_rules: etc/deprecated_rules.json
  packages: etc/packages.yml
  stack_schema_map: etc/stack-schema-map.yaml
  version_lock: etc/version.lock.json
rule_dirs:
- rules
- rules_building_block
testing:
  config: etc/example_test_config.yaml
```

```
package:
  name: '8.14'
```

```
8.14.0:
  beats: 8.12.2
  ecs: 8.11.0
  endgame: 8.4.0
```
</p>
</details> 

<details><summary>Testing</summary>
<p>

```
(detection-rules-build) ➜  detection-rules git:(3621-frdac-raise-a-better-exception-for-missing-contents-in-config-files) ✗ python -m detection_rules custom-rules setup-config mikas_folder 8.12
Loaded config file: /Users/stryker/workspace/Elastic/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Created directory: mikas_folder/actions
Created directory: mikas_folder/exceptions
Created directory: mikas_folder/rules
Created directory: mikas_folder/rules_building_block
Created directory: mikas_folder/etc
Created file with default content: mikas_folder/etc/deprecated_rules.json
Created file with default content: mikas_folder/etc/version.lock.json
Created file with default content: mikas_folder/etc/packages.yml
Created file with default content: mikas_folder/etc/stack-schema-map.yaml
Created file with default content: mikas_folder/etc/test_config.yaml
Created file with default content: mikas_folder/_config.yaml

# For details on how to configure the _config.yaml file,
# consult: /Users/stryker/workspace/Elastic/detection-rules/detection_rules/etc/_config.yaml
# or the docs: /Users/stryker/workspace/Elastic/detection-rules/docs/custom-rules.md
(detection-rules-build) ➜  detection-rules git:(3621-frdac-raise-a-better-exception-for-missing-contents-in-config-files) ✗ python -m detection_rules custom-rules setup-config mikas_folder     
Loaded config file: /Users/stryker/workspace/Elastic/detection-rules/.detection-rules-cfg.json

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Created directory: mikas_folder/actions
Created directory: mikas_folder/exceptions
Created directory: mikas_folder/rules
Created directory: mikas_folder/rules_building_block
Created directory: mikas_folder/etc
Created file with default content: mikas_folder/etc/deprecated_rules.json
Created file with default content: mikas_folder/etc/version.lock.json
Created file with default content: mikas_folder/etc/packages.yml
Created file with default content: mikas_folder/etc/stack-schema-map.yaml
Created file with default content: mikas_folder/etc/test_config.yaml
Created file with default content: mikas_folder/_config.yaml

# For details on how to configure the _config.yaml file,
# consult: /Users/stryker/workspace/Elastic/detection-rules/detection_rules/etc/_config.yaml
# or the docs: /Users/stryker/workspace/Elastic/detection-rules/docs/custom-rules.md
```

Make tests should still pass.

</p>
</details>

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,77 @@ def custom_rules():
         directory / 'exceptions',
         directory / 'rules',
-        etc_dir
+        directory / 'rules_building_block',
```
I don't think we need to do this - the goal has been to let users organize rules how they want, but I would always bury them under `rules/`

So if they _choose_ to do that, they can add `rules/building_block/`, etc.

```suggestion
```

Mikaayenson replied:
Then we should probably remove it from the example config.

eric-forte-elastic replied:
Are we also intending for a BBR directory to not be required in general then? At the moment I think that is is required, especially once we [decouple the default dependencies](https://github.com/elastic/detection-rules/issues/3619). If BBR is optional may impact how we want to implement some of the other DaC features.

brokensound77 replied:
I think it should be optional, or better, configurable. Keeping BBR in a separate folder is a construct _we_ have chosen to use, but by no means does it have to be a required thing

Mikaayenson replied:
I dont see an issue with leaving it in that case.

brokensound77 replied:
we should consider how to make it optional - but that should probably happen in your PR @eric-forte-elastic. Basically do a check first if the BBR path exists before calling `RuleCollection.default_bbr`

But also should update the docs here to reflect that this is _optional_ (probably some other doc updates required to about these changes overall)

eric-forte-elastic replied:
I agree that it should be optional, which I think then that would mean that we should leave it in here. 

This way we can make it "officially" optional with the relevant doc changes in subsequent PRs.

brokensound77 replied:
sgtm with doc updates

Mikaayenson replied:
Okay. So I'll disregard this.

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,77 @@ def custom_rules():
         etc_dir / 'stack-schema-map.yaml',
         etc_dir / 'version.lock.json',
-        etc_dir / 'test_config.yaml',
+        etc_dir / 'example_test_config.yaml',
```
the config is supposed to be called `test_config`. `example_test_config` is meant to be an example only and intentionally named different so as to _not_ be default loaded

brokensound77 replied:
```suggestion
        etc_dir / 'test_config.yaml',
```

brokensound77 commented on detection_rules/config.py:
```diff
@@ -156,10 +157,13 @@ def parse_rules_config(path: Optional[Path] = None) -> RulesConfig:
-        assert path.exists(), f'_config.yaml file missing in {CUSTOM_RULES_DIR}'
+    elif (path := Path(CUSTOM_RULES_DIR) / '_config.yaml').exists():
         loaded = yaml.safe_load(path.read_text())
+        if 'configuration_details' in loaded:
+            print(loaded['configuration_details'])
+            path.unlink()
+            sys.exit(0)
```
Can you explain the purpose of this

brokensound77 replied:
The only file that actually _requires_ data to not fail on default, is packages.yaml - explicitly the package version, which you could search for.

Additionally, you should not exit from here (especially with 0 code), but raise a specific error, like `ValueError`

Mikaayenson replied:
I'll update the summary.

brokensound77 replied:
Why do you delete the file? Also please raise an exception vs exiting

brokensound77 replied:
I would also recommend this:

>The only file that actually requires data to not fail on default, is packages.yaml - explicitly the package version, which you could search for.

over the search for `configuration_details` because if they create it themselves or dont use `--default` this will be missed

Mikaayenson replied:
> Why do you delete the file? Also please raise an exception vs exiting

I mentioned this already in the summary, but essentially, im thinking of an experience where users want defer to use the -use-default option after noticing the default doesn't get them much. With the way the `parse_rules_config` is called in so many different load times in the repo, you'll run into a blocker immediately. Removing the file after trying to use the config generated makes sure that you can call subsequent cli commands. It may be more helpful to run the commands in serial so you can see what I mean. Note: it doesn't remove the file immediately.

brokensound77 replied:
I don't think we should blindly delete files from users, especially based on assumptions. Better to raise a detailed error (missing data, refer to docs for file) and let them delete it.

Mikaayenson replied:
I was thinking about raising this to the CLI. Which I think I will.

Mikaayenson replied:
So the issue here is that the user doesn't even get a chance to specify by CLI to delete the file because it crashes by assert early in the execution. This time in rule.py

Mikaayenson replied:
I exposed deleting the file to the CLI.

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,77 @@ def custom_rules():
     """Commands for supporting custom rules."""
 
 
+def create_config_content(use_defaults, etc_dir):
```
type hints

brokensound77 replied:
return type

brokensound77 replied:
docstrings

brokensound77 commented on detection_rules/main.py:
```diff
@@ -419,7 +419,7 @@ def test_rules(ctx):
     rules_config = ctx.obj['rules_config']
     test_config = rules_config.test_config
-    tests, skipped = test_config.get_test_names(formatted=True)
+    skipped, tests = test_config.get_test_names(formatted=True)
```
Was this leftover from testing?

Mikaayenson replied:
I think its a bug introduced in your original PR

brokensound77 replied:
Can you share some example configs you are running that are inverted - I haven't seen this problem

Mikaayenson replied:
It's basically the default skipping one test. Instead of skipping one test, it skips all test except for the one you want to bypass

brokensound77 replied:
can you share the config for this scenario, the printed output, and your expected results

Mikaayenson replied:
[test.txt](https://github.com/elastic/detection-rules/files/15238604/test.txt)

Essentially I wanted to skip `- tests.test_all_rules.TestValidRules.test_schema_and_dupes` which makes sense because that test cases fails when running the test with no rules. 

Instead of bypassing the test case and running all others, it only tested the one I wanted to skip.

brokensound77 replied:
```suggestion
    tests, skipped = test_config.get_test_names(formatted=True)
```

Mikaayenson replied:
Ah so the issue was related to the way tests were being pulled `if self.unit_tests.test_only is not None`. This check essentially defaulted to an empty list since the config key was set without the values. I'll reset this and update this check.

brokensound77 commented:
In your testing output, can you also cat the `_config` file for both

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,77 @@ def custom_rules():
+            'version_lock': 'etc/version.lock.json',
+        },
+        'testing': {
+            'config': 'etc/example_test_config.yaml'
```
```suggestion
            'config': 'etc/test_config.yaml'
```

Mikaayenson commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,77 @@ def custom_rules():
+        }
+
+    click.echo(f'Configured _config.yaml with{" default contents" if use_defaults else " detailed instructions"}')
+    return yaml.safe_dump(config_content, default_flow_style=False)
```
I wanted to but didn't see one available.

eric-forte-elastic replied:
Agreed, looks like the potentially relevant utils functions are directory locked to `etc` so not useful here :+1: Thanks!

Mikaayenson commented:
> In your testing output, can you also cat the `_config` file for both

Updated the summary.

eric-forte-elastic reviewed (approved):
:green_circle: Manual review, looks good to me! :+1:

<details><summary>Testing :green_circle: </summary>
<p>

```shell
…ion-rules on  3621-frdac-raise-a-better-exception-for-missing-contents-in-config-files [$] is  v0.1.0 via  v3.12.3 (detection-rules-build) on  eric.forte 
❯ python -m detection_rules custom-rules init-config erics_folder

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Created directory: erics_folder/actions
Created directory: erics_folder/exceptions
Created directory: erics_folder/rules
Created directory: erics_folder/rules_building_block
Created directory: erics_folder/etc
Created file: erics_folder/etc/deprecated_rules.json
Created file: erics_folder/etc/packages.yml
Created file: erics_folder/etc/stack-schema-map.yaml
Created file: erics_folder/etc/version.lock.json
Created file: erics_folder/etc/test_config.yaml
Configured _config.yaml with detailed instructions

…on-rules on  3621-frdac-raise-a-better-exception-for-missing-contents-in-config-files [$?] is  v0.1.0 via  v3.12.3 (detection-rules-build) on  eric.forte 
❯ make test
Installing kql and kibana packages...
./env/detection-rules-build/bin/pip install lib/kql lib/kibana
Looking in indexes: https://pypi.org/simple, https://eric.forte%40elastic.co:****@artifactory.elastic.dev/artifactory/api/pypi/pypi-endgame/simple
Processing ./lib/kql
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Processing ./lib/kibana
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: eql==0.9.19 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection-rules-kql==0.1.7) (0.9.19)
Requirement already satisfied: lark-parser>=0.12.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection-rules-kql==0.1.7) (0.12.0)
Requirement already satisfied: requests<3.0,>=2.25 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection-rules-kibana==0.2.0) (2.31.0)
Requirement already satisfied: elasticsearch~=8.12.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection-rules-kibana==0.2.0) (8.12.1)
Requirement already satisfied: elastic-transport<9,>=8 in ./env/detection-rules-build/lib/python3.12/site-packages (from elasticsearch~=8.12.1->detection-rules-kibana==0.2.0) (8.12.0)
Requirement already satisfied: charset-normalizer<4,>=2 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests<3.0,>=2.25->detection-rules-kibana==0.2.0) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests<3.0,>=2.25->detection-rules-kibana==0.2.0) (3.6)
Requirement already satisfied: urllib3<3,>=1.21.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests<3.0,>=2.25->detection-rules-kibana==0.2.0) (2.2.1)
Requirement already satisfied: certifi>=2017.4.17 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests<3.0,>=2.25->detection-rules-kibana==0.2.0) (2024.2.2)
Building wheels for collected packages: detection-rules-kql, detection-rules-kibana
  Building wheel for detection-rules-kql (pyproject.toml) ... done
  Created wheel for detection-rules-kql: filename=detection_rules_kql-0.1.7-py3-none-any.whl size=16336 sha256=b16342078afe841fa8e9d5a956d41610a7a84bf7ebe337b72b1094b56634b59f
  Stored in directory: /tmp/pip-ephem-wheel-cache-aq1z54ig/wheels/56/d8/16/54016845ffe39458d941c5c2dbc03e4998264eeec10328eebb
  Building wheel for detection-rules-kibana (pyproject.toml) ... done
  Created wheel for detection-rules-kibana: filename=detection_rules_kibana-0.2.0-py3-none-any.whl size=9325 sha256=775ecf84396d2ffce3a8815fe8ed048bbe0b2b23c2ea9575f1e7a2c772406fcd
  Stored in directory: /tmp/pip-ephem-wheel-cache-aq1z54ig/wheels/05/3b/89/3dab4aeb78501459dd86df8d126afb94aab37d28aea1263c1d
Successfully built detection-rules-kql detection-rules-kibana
Installing collected packages: detection-rules-kql, detection-rules-kibana
  Attempting uninstall: detection-rules-kql
    Found existing installation: detection-rules-kql 0.1.8
    Uninstalling detection-rules-kql-0.1.8:
      Successfully uninstalled detection-rules-kql-0.1.8
  Attempting uninstall: detection-rules-kibana
    Found existing installation: detection-rules-kibana 0.1.0
    Uninstalling detection-rules-kibana-0.1.0:
      Successfully uninstalled detection-rules-kibana-0.1.0
Successfully installed detection-rules-kibana-0.2.0 detection-rules-kql-0.1.7
Installing all dependencies...
./env/detection-rules-build/bin/pip install .[dev]
Looking in indexes: https://pypi.org/simple, https://eric.forte%40elastic.co:****@artifactory.elastic.dev/artifactory/api/pypi/pypi-endgame/simple
Processing /home/forteea1/Code/clean_mains/detection-rules
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Requirement already satisfied: Click~=8.1.7 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (8.1.7)
Requirement already satisfied: elasticsearch~=8.12.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (8.12.1)
Requirement already satisfied: eql==0.9.19 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.9.19)
Requirement already satisfied: jsl==0.2.4 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.2.4)
Requirement already satisfied: jsonschema>=4.21.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (4.21.1)
Requirement already satisfied: marko==2.0.3 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (2.0.3)
Requirement already satisfied: marshmallow-dataclass~=8.6.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from marshmallow-dataclass[union]~=8.6.0->detection_rules==0.1.0) (8.6.1)
Requirement already satisfied: marshmallow-jsonschema~=0.13.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.13.0)
Requirement already satisfied: marshmallow-union~=0.1.15 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.1.15.post1)
Requirement already satisfied: marshmallow~=3.21.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (3.21.1)
Requirement already satisfied: pytoml==0.1.21 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.1.21)
Requirement already satisfied: PyYAML~=6.0.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (6.0.1)
Requirement already satisfied: requests~=2.31.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (2.31.0)
Requirement already satisfied: toml==0.10.2 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.10.2)
Requirement already satisfied: typing-inspect==0.9.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.9.0)
Requirement already satisfied: typing-extensions==4.10.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (4.10.0)
Requirement already satisfied: XlsxWriter~=3.2.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (3.2.0)
Requirement already satisfied: semver==3.0.2 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (3.0.2)
Requirement already satisfied: lark-parser~=0.12.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from eql==0.9.19->detection_rules==0.1.0) (0.12.0)
Requirement already satisfied: mypy-extensions>=0.3.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from typing-inspect==0.9.0->detection_rules==0.1.0) (1.0.0)
Requirement already satisfied: pep8-naming==0.13.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (0.13.0)
Requirement already satisfied: PyGithub==2.2.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (2.2.0)
Requirement already satisfied: flake8==7.0.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (7.0.0)
Requirement already satisfied: pyflakes==3.2.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (3.2.0)
Requirement already satisfied: pytest>=8.1.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (8.1.1)
Requirement already satisfied: pre-commit==3.6.2 in ./env/detection-rules-build/lib/python3.12/site-packages (from detection_rules==0.1.0) (3.6.2)
Requirement already satisfied: mccabe<0.8.0,>=0.7.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from flake8==7.0.0->detection_rules==0.1.0) (0.7.0)
Requirement already satisfied: pycodestyle<2.12.0,>=2.11.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from flake8==7.0.0->detection_rules==0.1.0) (2.11.1)
Requirement already satisfied: cfgv>=2.0.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from pre-commit==3.6.2->detection_rules==0.1.0) (3.4.0)
Requirement already satisfied: identify>=1.0.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from pre-commit==3.6.2->detection_rules==0.1.0) (2.5.35)
Requirement already satisfied: nodeenv>=0.11.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from pre-commit==3.6.2->detection_rules==0.1.0) (1.8.0)
Requirement already satisfied: virtualenv>=20.10.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from pre-commit==3.6.2->detection_rules==0.1.0) (20.25.1)
Requirement already satisfied: pynacl>=1.4.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from PyGithub==2.2.0->detection_rules==0.1.0) (1.5.0)
Requirement already satisfied: pyjwt>=2.4.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from pyjwt[crypto]>=2.4.0->PyGithub==2.2.0->detection_rules==0.1.0) (2.8.0)
Requirement already satisfied: urllib3>=1.26.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from PyGithub==2.2.0->detection_rules==0.1.0) (2.2.1)
Requirement already satisfied: Deprecated in ./env/detection-rules-build/lib/python3.12/site-packages (from PyGithub==2.2.0->detection_rules==0.1.0) (1.2.14)
Requirement already satisfied: elastic-transport<9,>=8 in ./env/detection-rules-build/lib/python3.12/site-packages (from elasticsearch~=8.12.1->detection_rules==0.1.0) (8.12.0)
Requirement already satisfied: attrs>=22.2.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from jsonschema>=4.21.1->detection_rules==0.1.0) (23.2.0)
Requirement already satisfied: jsonschema-specifications>=2023.03.6 in ./env/detection-rules-build/lib/python3.12/site-packages (from jsonschema>=4.21.1->detection_rules==0.1.0) (2023.12.1)
Requirement already satisfied: referencing>=0.28.4 in ./env/detection-rules-build/lib/python3.12/site-packages (from jsonschema>=4.21.1->detection_rules==0.1.0) (0.33.0)
Requirement already satisfied: rpds-py>=0.7.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from jsonschema>=4.21.1->detection_rules==0.1.0) (0.18.0)
Requirement already satisfied: packaging>=17.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from marshmallow~=3.21.1->detection_rules==0.1.0) (24.0)
Requirement already satisfied: typeguard<4.0.0,>=2.4.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from marshmallow-dataclass[union]~=8.6.0->detection_rules==0.1.0) (3.0.2)
Requirement already satisfied: iniconfig in ./env/detection-rules-build/lib/python3.12/site-packages (from pytest>=8.1.1->detection_rules==0.1.0) (2.0.0)
Requirement already satisfied: pluggy<2.0,>=1.4 in ./env/detection-rules-build/lib/python3.12/site-packages (from pytest>=8.1.1->detection_rules==0.1.0) (1.4.0)
Requirement already satisfied: charset-normalizer<4,>=2 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests~=2.31.0->detection_rules==0.1.0) (3.3.2)
Requirement already satisfied: idna<4,>=2.5 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests~=2.31.0->detection_rules==0.1.0) (3.6)
Requirement already satisfied: certifi>=2017.4.17 in ./env/detection-rules-build/lib/python3.12/site-packages (from requests~=2.31.0->detection_rules==0.1.0) (2024.2.2)
Requirement already satisfied: setuptools in ./env/detection-rules-build/lib/python3.12/site-packages (from nodeenv>=0.11.1->pre-commit==3.6.2->detection_rules==0.1.0) (69.2.0)
Requirement already satisfied: cryptography>=3.4.0 in ./env/detection-rules-build/lib/python3.12/site-packages (from pyjwt[crypto]>=2.4.0->PyGithub==2.2.0->detection_rules==0.1.0) (42.0.5)
Requirement already satisfied: cffi>=1.4.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from pynacl>=1.4.0->PyGithub==2.2.0->detection_rules==0.1.0) (1.16.0)
Requirement already satisfied: distlib<1,>=0.3.7 in ./env/detection-rules-build/lib/python3.12/site-packages (from virtualenv>=20.10.0->pre-commit==3.6.2->detection_rules==0.1.0) (0.3.8)
Requirement already satisfied: filelock<4,>=3.12.2 in ./env/detection-rules-build/lib/python3.12/site-packages (from virtualenv>=20.10.0->pre-commit==3.6.2->detection_rules==0.1.0) (3.13.1)
Requirement already satisfied: platformdirs<5,>=3.9.1 in ./env/detection-rules-build/lib/python3.12/site-packages (from virtualenv>=20.10.0->pre-commit==3.6.2->detection_rules==0.1.0) (4.2.0)
Requirement already satisfied: wrapt<2,>=1.10 in ./env/detection-rules-build/lib/python3.12/site-packages (from Deprecated->PyGithub==2.2.0->detection_rules==0.1.0) (1.16.0)
Requirement already satisfied: pycparser in ./env/detection-rules-build/lib/python3.12/site-packages (from cffi>=1.4.1->pynacl>=1.4.0->PyGithub==2.2.0->detection_rules==0.1.0) (2.21)
Building wheels for collected packages: detection_rules
  Building wheel for detection_rules (pyproject.toml) ... done
  Created wheel for detection_rules: filename=detection_rules-0.1.0-py3-none-any.whl size=39609775 sha256=0f2e6ebc4e628f20521bf2adc0e300f7e888490489e0f8fdd6fa896e97edec08
  Stored in directory: /home/forteea1/.cache/pip/wheels/33/0b/6f/442542fc0e808e368b4dfaf768ed0b61a5d8281942974600d4
Successfully built detection_rules
Installing collected packages: detection_rules
  Attempting uninstall: detection_rules
    Found existing installation: detection_rules 0.1.0
    Uninstalling detection_rules-0.1.0:
      Successfully uninstalled detection_rules-0.1.0
Successfully installed detection_rules-0.1.0
LINTING
./env/detection-rules-build/bin/python -m flake8 tests detection_rules --ignore D203,N815 --max-line-length 120
./env/detection-rules-build/bin/python -m detection_rules test

█▀▀▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄▄▄ ▄   ▄      █▀▀▄ ▄  ▄ ▄   ▄▄▄ ▄▄▄
█  █ █▄▄  █  █▄▄ █    █   █  █ █ █▀▄ █      █▄▄▀ █  █ █   █▄▄ █▄▄
█▄▄▀ █▄▄  █  █▄▄ █▄▄  █  ▄█▄ █▄█ █ ▀▄█      █ ▀▄ █▄▄█ █▄▄ █▄▄ ▄▄█

Tests skipped per config (144):
tests/kuery/test_dsl.py::TestKQLtoDSL::test_and_query
tests/kuery/test_dsl.py::TestKQLtoDSL::test_field_exists
tests/kuery/test_dsl.py::TestKQLtoDSL::test_field_inequality
tests/kuery/test_dsl.py::TestKQLtoDSL::test_field_match
tests/kuery/test_dsl.py::TestKQLtoDSL::test_not_query
tests/kuery/test_dsl.py::TestKQLtoDSL::test_optimizations
tests/kuery/test_dsl.py::TestKQLtoDSL::test_or_query
tests/kuery/test_eql2kql.py::TestEql2Kql::test_and_query
tests/kuery/test_eql2kql.py::TestEql2Kql::test_boolean_precedence
tests/kuery/test_eql2kql.py::TestEql2Kql::test_field_equals
tests/kuery/test_eql2kql.py::TestEql2Kql::test_field_inequality
tests/kuery/test_eql2kql.py::TestEql2Kql::test_ip_checks
tests/kuery/test_eql2kql.py::TestEql2Kql::test_list_of_values
tests/kuery/test_eql2kql.py::TestEql2Kql::test_not_query
tests/kuery/test_eql2kql.py::TestEql2Kql::test_or_query
tests/kuery/test_eql2kql.py::TestEql2Kql::test_wildcard_field
tests/kuery/test_evaluator.py::EvaluatorTests::test_and_expr
tests/kuery/test_evaluator.py::EvaluatorTests::test_and_values
tests/kuery/test_evaluator.py::EvaluatorTests::test_cidr_match
tests/kuery/test_evaluator.py::EvaluatorTests::test_field_exists
tests/kuery/test_evaluator.py::EvaluatorTests::test_flattening
tests/kuery/test_evaluator.py::EvaluatorTests::test_list_value
tests/kuery/test_evaluator.py::EvaluatorTests::test_not_value
tests/kuery/test_evaluator.py::EvaluatorTests::test_or_expr
tests/kuery/test_evaluator.py::EvaluatorTests::test_or_values
tests/kuery/test_evaluator.py::EvaluatorTests::test_quoted_wildcard
tests/kuery/test_evaluator.py::EvaluatorTests::test_range
tests/kuery/test_evaluator.py::EvaluatorTests::test_single_value
tests/kuery/test_evaluator.py::EvaluatorTests::test_wildcard
tests/kuery/test_kql2eql.py::TestKql2Eql::test_and_query
tests/kuery/test_kql2eql.py::TestKql2Eql::test_boolean_precedence
tests/kuery/test_kql2eql.py::TestKql2Eql::test_field_equals
tests/kuery/test_kql2eql.py::TestKql2Eql::test_field_inequality
tests/kuery/test_kql2eql.py::TestKql2Eql::test_list_of_values
tests/kuery/test_kql2eql.py::TestKql2Eql::test_lone_value
tests/kuery/test_kql2eql.py::TestKql2Eql::test_nested_query
tests/kuery/test_kql2eql.py::TestKql2Eql::test_not_query
tests/kuery/test_kql2eql.py::TestKql2Eql::test_or_query
tests/kuery/test_kql2eql.py::TestKql2Eql::test_schema
tests/kuery/test_lint.py::LintTests::test_and_not
tests/kuery/test_lint.py::LintTests::test_compound
tests/kuery/test_lint.py::LintTests::test_double_negate
tests/kuery/test_lint.py::LintTests::test_extract_not
tests/kuery/test_lint.py::LintTests::test_ip
tests/kuery/test_lint.py::LintTests::test_lint_field
tests/kuery/test_lint.py::LintTests::test_lint_precedence
tests/kuery/test_lint.py::LintTests::test_merge_fields
tests/kuery/test_lint.py::LintTests::test_mixed_demorgans
tests/kuery/test_lint.py::LintTests::test_not_demorgans
tests/kuery/test_lint.py::LintTests::test_not_or
tests/kuery/test_lint.py::LintTests::test_upper_tokens
tests/kuery/test_parser.py::ParserTests::test_conversion
tests/kuery/test_parser.py::ParserTests::test_date
tests/kuery/test_parser.py::ParserTests::test_keyword
tests/kuery/test_parser.py::ParserTests::test_list_equals
tests/kuery/test_parser.py::ParserTests::test_multiple_types_fail
tests/kuery/test_parser.py::ParserTests::test_multiple_types_success
tests/kuery/test_parser.py::ParserTests::test_number_exists
tests/kuery/test_parser.py::ParserTests::test_number_wildcard_fail
tests/kuery/test_parser.py::ParserTests::test_type_family_fail
tests/kuery/test_parser.py::ParserTests::test_type_family_success
tests/test_all_rules.py::TestAlertSuppression::test_group_field_in_schemas
tests/test_all_rules.py::TestBuildTimeFields::test_build_fields_min_stack
tests/test_all_rules.py::TestIncompatibleFields::test_rule_backports_for_restricted_fields
tests/test_all_rules.py::TestIntegrationRules::test_all_min_stack_rules_have_comment
tests/test_all_rules.py::TestIntegrationRules::test_integration_guide
tests/test_all_rules.py::TestIntegrationRules::test_ml_integration_jobs_exist
tests/test_all_rules.py::TestIntegrationRules::test_rule_demotions
tests/test_all_rules.py::TestLicense::test_elastic_license_only_v2
tests/test_all_rules.py::TestNoteMarkdownPlugins::test_if_plugins_explicitly_defined
tests/test_all_rules.py::TestNoteMarkdownPlugins::test_note_has_osquery_warning
tests/test_all_rules.py::TestNoteMarkdownPlugins::test_plugin_placeholders_match_entries
tests/test_all_rules.py::TestRiskScoreMismatch::test_rule_risk_score_severity_mismatch
tests/test_all_rules.py::TestRuleFiles::test_bbr_in_correct_dir
tests/test_all_rules.py::TestRuleFiles::test_non_bbr_in_correct_dir
tests/test_all_rules.py::TestRuleFiles::test_rule_file_name_tactic
tests/test_all_rules.py::TestRuleMetadata::test_deprecated_rules
tests/test_all_rules.py::TestRuleMetadata::test_event_dataset
tests/test_all_rules.py::TestRuleMetadata::test_integration_tag
tests/test_all_rules.py::TestRuleMetadata::test_invalid_queries
tests/test_all_rules.py::TestRuleMetadata::test_updated_date_newer_than_creation
tests/test_all_rules.py::TestRuleTags::test_casing_and_spacing
tests/test_all_rules.py::TestRuleTags::test_investigation_guide_tag
tests/test_all_rules.py::TestRuleTags::test_ml_rule_type_tags
tests/test_all_rules.py::TestRuleTags::test_no_duplicate_tags
tests/test_all_rules.py::TestRuleTags::test_os_tags
tests/test_all_rules.py::TestRuleTags::test_primary_tactic_as_tag
tests/test_all_rules.py::TestRuleTags::test_required_tags
tests/test_all_rules.py::TestRuleTags::test_tag_prefix
tests/test_all_rules.py::TestRuleTimelines::test_timeline_has_title
tests/test_all_rules.py::TestRuleTiming::test_eql_interval_to_maxspan
tests/test_all_rules.py::TestRuleTiming::test_eql_lookback
tests/test_all_rules.py::TestRuleTiming::test_event_override
tests/test_all_rules.py::TestRuleTiming::test_required_lookback
tests/test_all_rules.py::TestThreatMappings::test_duplicated_tactics
tests/test_all_rules.py::TestThreatMappings::test_tactic_to_technique_correlations
tests/test_all_rules.py::TestThreatMappings::test_technique_deprecations
tests/test_all_rules.py::TestValidRules::test_all_rule_queries_optimized
tests/test_all_rules.py::TestValidRules::test_bbr_validation
tests/test_all_rules.py::TestValidRules::test_duplicate_file_names
tests/test_all_rules.py::TestValidRules::test_file_names
tests/test_all_rules.py::TestValidRules::test_production_rules_have_rta
tests/test_all_rules.py::TestValidRules::test_rule_type_changes
tests/test_all_rules.py::TestValidRules::test_schema_and_dupes
tests/test_gh_workflows.py::TestWorkflows::test_matrix_to_lock_version_defaults
tests/test_hunt_data.py::TestHunt::test_load_toml_files
tests/test_hunt_data.py::TestHunt::test_markdown_existence
tests/test_hunt_data.py::TestHunt::test_toml_loading
tests/test_mappings.py::TestMappings::test_false_positives
tests/test_mappings.py::TestMappings::test_true_positives
tests/test_mappings.py::TestRTAs::test_rtas_with_triggered_rules_have_uuid
tests/test_packages.py::TestPackages::test_package_loader_default_configs
tests/test_packages.py::TestPackages::test_package_loader_production_config
tests/test_packages.py::TestPackages::test_package_summary
tests/test_packages.py::TestPackages::test_rule_versioning
tests/test_packages.py::TestRegistryPackage::test_registry_package_config
tests/test_python_library.py::TestEQLInSet::test_eql_in_set
tests/test_schemas.py::TestSchemas::test_eql_validation
tests/test_schemas.py::TestSchemas::test_query_downgrade_7_x
tests/test_schemas.py::TestSchemas::test_query_downgrade_8_x
tests/test_schemas.py::TestSchemas::test_threshold_downgrade_7_x
tests/test_schemas.py::TestSchemas::test_threshold_downgrade_8_x
tests/test_schemas.py::TestSchemas::test_versioned_downgrade_7_x
tests/test_schemas.py::TestSchemas::test_versioned_downgrade_8_x
tests/test_schemas.py::TestVersionLockSchema::test_version_lock_has_nested_previous
tests/test_schemas.py::TestVersionLockSchema::test_version_lock_no_previous
tests/test_schemas.py::TestVersions::test_stack_schema_map
tests/test_specific_rules.py::TestESQLRules::test_esql_queries
tests/test_specific_rules.py::TestEndpointQuery::test_os_and_platform_in_query
tests/test_specific_rules.py::TestNewTerms::test_history_window_start
tests/test_specific_rules.py::TestNewTerms::test_new_terms_field_exists
tests/test_specific_rules.py::TestNewTerms::test_new_terms_fields
tests/test_specific_rules.py::TestNewTerms::test_new_terms_fields_unique
tests/test_specific_rules.py::TestNewTerms::test_new_terms_max_limit
tests/test_toml_formatter.py::TestRuleTomlFormatter::test_formatter_deep
tests/test_toml_formatter.py::TestRuleTomlFormatter::test_formatter_rule
tests/test_toml_formatter.py::TestRuleTomlFormatter::test_normalization
tests/test_transform_fields.py::TestGuideMarkdownPlugins::test_plugin_conversion
tests/test_transform_fields.py::TestGuideMarkdownPlugins::test_transform_guide_markdown_plugins
tests/test_utils.py::TestTimeUtils::test_caching
tests/test_utils.py::TestTimeUtils::test_event_class_normalization
tests/test_utils.py::TestTimeUtils::test_schema_multifields
tests/test_utils.py::TestTimeUtils::test_time_normalize
tests/test_version_locking.py::TestVersionLock::test_previous_entries_gte_current_min_stack
No tests found to execute!


```

</p>
</details>

Mikaayenson commented on detection_rules/config.py:
```diff
@@ -156,10 +157,13 @@ def parse_rules_config(path: Optional[Path] = None) -> RulesConfig:
+        if 'configuration_details' in loaded:
+            print(loaded['configuration_details'])
+            path.unlink()
+            sys.exit(0)
```
It's probably worth noting that if the config was implemented with a lazy load option, it probably would save us some trouble.

brokensound77 replied:
>Not showing an entire stack trace which can be misinterpreted as broken code by unexperienced users.

If this is your concern, catch it from the caller (CLI) and print a more specific, slimmed down error (or even `ClientError`)

Mikaayenson replied:
I think the only issue like I mentioned is that the existing logic calls `parse_rules_config` as a global variable in some init files, which I thought occurred earlier in execution.

eric-forte-elastic replied:
While I understand the hesitation of using sys.exit for handling errors, it is a clean way (and my understanding it is the preferred clean way) of exiting a python program where there is no error. Also, for the case of preferring ctx.exit I agree, when this is available we should use it; however, I do not think we would yet be in the context of click. 

As it is written I think the `parse_rules_config` would load during the imports prior to the establishing of the click context, making for either confusing code or potentially a failed call. It may appear to function but the desired behavior of gracefully terminating from the context may not be available. 

For a graceful, non-error termination sys.exit wraps a call to `raise SystemExit`. While we could directly use `raise SystemExit` this could also be confusing as raise is typically reserved for exceptions and this would imply that we are raising due to an error, which we are not. Functionally, `raise SystemExit` and sys.exit are in effect identical, so that is more of a stylistic preference, but I would err on the side of being clear that this is not an error.

brokensound77 replied:
we do not track specific error codes to types of errors. An exception will result in python exiting with a non-zero, achieving the desired result. Wrapping in `sys.exit` does nothing but hide the stack trace as you intended, however, I think that is the better thing to do here - let is raise. The error string is obvious enough, and it points users to where in the code it is erroring if they want to understand.

Otherwise, we can force the parse_config to load predictably (since we _can_ lazy load it) and wrap it in an exception there

Mikaayenson replied:
If we intend to lazy load, let's do it outside of this PR. Outside of making adding exception handling to `parse_rules_config` calls, im more inclined to leave this as-is.

Mikaayenson replied:
> it looks like it is `detection-rules/schemas/__init__.py`
> 
> if we add this to `__main__` we can reliably control the error
> 
> ```python
> try:
>     RULES_CONFIG = parse_rules_config()
> except SpecificErrorRaised as e:
>     print(pretty error, sys.stderr)
>     raise # (or sys.exit)
> ```

Let's do that. Not the best, but IMO better than the alternatives.

brokensound77 replied:
based on how the repo operates, we really _can't_ lazy load. I think this is the best way to do it. The question still remains though - raise exception or print and sys.exit

Mikaayenson replied:
We could call try/except everywhere that uses calls this function when setting a global variable, but it really starts looking messy.

Mikaayenson replied:
![Screenshot 2024-05-07 at 4 50 59 PM](https://github.com/elastic/detection-rules/assets/1636709/95eed14b-5f7a-4297-ae1e-8b744bfee557)

Unfortunately this was added like this in 8 different places so im not sure if try/except is the way to go.

Mikaayenson replied:
The only other thing remaining 3653 i think is this. TLDR; assert with stack trace, or exit clean with sys.exit. IMO adding try/excepts around RULES_CONFIG = parse_rules_config() throughout the 8 global variables in the repo isn't the best solution. If there are no other ideas, I'm inclined to leave the `sys.exit` line in, since it potentially provides a cleaner experience for users.

eric-forte-elastic replied:
I agree :+1:  I think that the cleanest method of dealing with this situation is to use `sys.exit` with the `0` argument (or no argument at all as the default is `0`).

Mikaayenson replied:
🚀 Using @eric-forte-elastic [approach](https://github.com/elastic/detection-rules/pull/3653#discussion_r1593092457), I think is a happy medium so I rolled that in!

brokensound77 commented on detection_rules/config.py:
```diff
@@ -100,7 +101,7 @@ def get_test_names(self, formatted: bool = False) -> (List[str], List[str]):
 
         combined_tests = sorted(set(defined_tests + self.tests_by_patterns(*patterns)))
 
-        if self.unit_tests.test_only is not None:
```
revert this and update `UnitTest.__post_init__` assertion error to

```python
        assert (self.bypass is not None and self.test_only is not None), 'Cannot set both `test_only` and `bypass` in test_config!'
```

Mikaayenson replied:
Strange duplicate.

brokensound77 commented on detection_rules/config.py:
```diff
@@ -100,7 +101,7 @@ def get_test_names(self, formatted: bool = False) -> (List[str], List[str]):
 
         combined_tests = sorted(set(defined_tests + self.tests_by_patterns(*patterns)))
 
-        if self.unit_tests.test_only is not None:
```
revert this and update `UnitTest.__post_init__` assertion error to

```python
        assert (self.bypass is not None and self.test_only is not None), 'Cannot set both `test_only` and `bypass` in test_config!'
```

Mikaayenson replied:
im not sure if this is exactly right, but I think I get the idea.

Mikaayenson replied:
![Screenshot 2024-05-07 at 4 33 06 PM](https://github.com/elastic/detection-rules/assets/1636709/a8ef242a-16f8-44e0-a595-cfd5e8346004)

Mikaayenson replied:
Oh I see. I used the slack version. I see its updated.

Mikaayenson commented on detection_rules/config.py:
```diff
@@ -150,16 +151,39 @@ class RulesConfig:
+    elif CUSTOM_RULES_DIR and (path := Path(CUSTOM_RULES_DIR) / '_config.yaml').exists():
         loaded = yaml.safe_load(path.read_text())
+        if 'configuration_details' in loaded or not loaded:
+            print(loaded['configuration_details'])
+            raise SystemExit(0)
```
We can also combine these two lines if we really think this is an error. IMO it's more of a NOP than an error, hence the `0`.

eric-forte-elastic replied:
Sounds good to me, same functionality as `sys.exit` :+1:

brokensound77 commented on detection_rules/config.py:
```diff
@@ -100,7 +101,7 @@ def get_test_names(self, formatted: bool = False) -> (List[str], List[str]):
         combined_tests = sorted(set(defined_tests + self.tests_by_patterns(*patterns)))
 
-        if self.unit_tests.test_only is not None:
+        if self.unit_tests.test_only:
```
can you revert this (the assertion error above was the fix - this is still needed)

```suggestion
        if self.unit_tests.test_only is not None:
```

brokensound77 reviewed (changes requested):
Let's discuss the sys.exit vs exception before proceeding

brokensound77 commented on detection_rules/config.py:
```diff
@@ -150,16 +151,38 @@ class RulesConfig:
+
+    paths = [
+        ('name',),
+        ('registry_data', 'conditions'),
```
nothing under registry data should be required (if so we need to decouple) - that is only relevant for EPR fleet releasing.

Mikaayenson replied:
https://github.com/elastic/detection-rules/pull/3653#discussion_r1594303263

brokensound77 commented on detection_rules/config.py:
```diff
@@ -150,16 +151,38 @@ class RulesConfig:
+
+    paths = [
+        ('name',),
+        ('registry_data', 'conditions'),
```
nothing under registry data should be required (if so we need to decouple) - that is only relevant for EPR fleet releasing.

IINM, the only _truly_ required is `package.name`

Mikaayenson replied:
https://github.com/elastic/detection-rules/pull/3653#discussion_r1594303263

brokensound77 commented on detection_rules/config.py:
```diff
@@ -196,6 +219,11 @@ def parse_rules_config(path: Optional[Path] = None) -> RulesConfig:
     contents = {k: load_dump(str(base_dir.joinpath(v))) for k, v in loaded['files'].items()}
+
+    # check required package fields
+    if 'packages' in contents and 'package' in contents['packages']:
```
based on this [comment](https://github.com/elastic/detection-rules/pull/3653/files#r1594288060), you can probably delete `check_path` and `validate_package_contents` and just do the check for `package.name`

Mikaayenson replied:
If we want to support people managing schemas then they need the `kibana.version`.

brokensound77 replied:
name is version

Mikaayenson replied:
Oh right! so maybe we can simplify this.

brokensound77 commented on detection_rules/config.py:
```diff
@@ -150,16 +151,38 @@ class RulesConfig:
+    elif CUSTOM_RULES_DIR and (path := Path(CUSTOM_RULES_DIR) / '_config.yaml').exists():
         loaded = yaml.safe_load(path.read_text())
+        if 'configuration_details' in loaded or not loaded:
+            raise SystemExit(loaded['configuration_details'])
```
```suggestion
            raise ValueError(f'Missing required content for config file: {path} (default contents detected')
```

This stays consistent with raising a `ValueError` for missing contents (should be the same).

Then in `__main__.py`

```python
try:
    RULES_CONFIG = parse_rules_config()
except SpecificErrorRaised as e:
    print('There was an error loading the _config.yml file', file=sys.stderr)  # or click.echo
    sys.exit(not 0) # or raise
```

Mikaayenson replied:
Based on the discussion in the original [thread](https://github.com/elastic/detection-rules/pull/3653#discussion_r1594157686). We will have to make this change in multiple changes, which IMO is not the best/cleanest approach. So to meet in the middle, I'm raising a SystemExit vs (sys.exit or the try except approach you mentioned).

brokensound77 replied:
no you dont - main is the first load and the whole point of putting it there. It forces an early load and an early failure - the rest are _cached_ loads

brokensound77 replied:
also rather than checking for `configuration_details` here, it would be more resilient to check for an absence of the required keys

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,102 @@ def custom_rules():
+@click.option('--use-defaults', is_flag=True, help="Use default contents from detection_rules/etc folder.")
+@click.option('--delete-config', is_flag=True, help="Delete the existing _config.yaml file.")
+def init_config(directory: Path, use_defaults: bool, delete_config: bool):
     """Initialize the custom rules configuration."""
```
```suggestion
    """Initialize the custom rules configuration directory and files."""
```

Mikaayenson commented on detection_rules/custom_rules.py:
```diff
@@ -20,30 +20,102 @@ def custom_rules():
 @click.argument('directory', type=Path)
-def init_config(directory: Path):
+@click.option('--use-defaults', is_flag=True, help="Use default contents from detection_rules/etc folder.")
+@click.option('--delete-config', is_flag=True, help="Delete the existing _config.yaml file.")
```
skipping on the prompt

Mikaayenson commented:
Alright! So after internal discussion here was what I implemented based on what we all agreed on.

- `init-config` --> `setup-config`: Not just base file structure, but setup with initial file content so users can easily get to testing, and removed the config details since we're not defaulting to initialize files with content. Added a cli param to allow users to specify kibana version during setup. Also echoed doc links post setup.
- `SystemExit`: Decided to keep this raised exit since the `RULES_CONFIG = parse_rules_config()` is initialized in so many places in the repo. Also because we dont have CLI context to try/catch.
- Moved `_config.yml` validation to a dataclass (note: couldn't use mixins because of the multiple circular dependancies)
- Added `packages.yaml` validation to the post init of the TestConfig

LMK if I missed anything!

Mikaayenson commented on detection_rules/config.py:
```diff
@@ -41,6 +42,44 @@ def __post_init__(self):
+    testing: Optional[TestConfigPath] = None
+
+    @classmethod
+    def from_dict(cls, obj: dict):
```
usually would add a docstring, but our other `from_dict` methods dont.

brokensound77 commented on detection_rules/config.py:
```diff
@@ -149,6 +188,14 @@ class RulesConfig:
 
+    def __post_init__(self):
+        """Perform post validation on packages.yml file."""
+        if 'package' not in self.packages:
+            raise SystemExit('Missing the `package` field defined in packages.yml.')
+
+        if 'name' not in self.packages['package']:
+            raise SystemExit('Missing the `name` field defined in packages.yml.')
```
I would raise `ValueError` here and then below, try/except and catch it and _then_ raise `SystemExit` from there

```suggestion
        if 'package' not in self.packages and  'name' not in self.packages['package']:
            raise ValueError('Missing required `package.name` value within: {self.packages_file}.')
```

Mikaayenson replied:
makes sense!

brokensound77 commented on detection_rules/config.py:
```diff
@@ -158,13 +205,17 @@ def parse_rules_config(path: Optional[Path] = None) -> RulesConfig:
+    except KeyError as e:
+        raise SystemExit(f'Missing key `{str(e)}` in _config.yaml file.')
+    except TypeError:
+        raise SystemExit(f'No data loaded from {path}')
```
```suggestion
        raise SystemExit(f'No data loaded from {path}')
    except ValueError as e:
        raise SystemExit(e)
```

Mikaayenson replied:
We probably don't need this. Might be good to add just in case. If this is related to `RulesConfig`, remember the initialization happens later.

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -7,43 +7,120 @@
+def setup_config(directory: Path, kibana_version: str, delete_config: bool):
+    """Setup the custom rules configuration directory and files with defaults."""
+
+    if delete_config:
+        config = directory / '_config.yaml'
+        if config.exists():
+            config.unlink()
+            click.echo(f'Deleted existing _config.yaml file: {config}')
+
```
the way this is implemented is more of an _overwrite_ than _delete_, since the file is going to be written anyway. Thoughts on this?

```suggestion
    config = directory / '_config.yaml'
    if not overwrite and config.exists():
        raise FileExistsError(f'{config} already exists. Use --overwrite to update')
```

(then you can remove config from L71)

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -7,43 +7,120 @@
-def init_config(directory: Path):
-    """Initialize the custom rules configuration."""
+@click.argument('kibana-version', type=str, default=load_etc_dump('packages.yml')['package']['name'])
+@click.option('--delete-config', is_flag=True, help="Delete the existing _config.yaml file.")
```
```suggestion
@click.option('--overwrite', is_flag=True, help="Overwrite the existing _config.yaml file.")
```

brokensound77 reviewed (commented):
Thanks for the changes - we're almost there; just a few remaining comments

Mikaayenson commented:
I'd like to get @traut to put final 👀 on this but if it's a blocker to your PR, I can merge it in today @eric-forte-elastic.

eric-forte-elastic commented:
> I'd like to get @traut to put final 👀 on this but if it's a blocker to your PR, I can merge it in today @eric-forte-elastic.

I don't think its too much of a blocker to need to merge without @traut looking at it. I've just been merging in your changes to this branch into mine, so I can do the same thing, especially if we do not expect it to change much now,. We will just need to merge this in before merging my PR.

brokensound77 commented on detection_rules/config.py:
```diff
@@ -41,6 +42,44 @@ def __post_init__(self):
+    testing: Optional[TestConfigPath] = None
+
+    @classmethod
+    def from_dict(cls, obj: dict):
```
```suggestion
    def from_dict(cls, obj: dict) -> ConfigFile:
```

Mikaayenson replied:
its not instantiated at this point, but we can still add some hinting.

brokensound77 commented on detection_rules/custom_rules.py:
```diff
@@ -7,43 +7,118 @@
+        raise FileExistsError(f'{config} already exists. Use --overwrite to update')
+
     etc_dir = directory / 'etc'
     config = directory / '_config.yaml'
```
```suggestion
```

brokensound77 reviewed (approved):
a few notes, otherwise LGTM 🚀

traut commented on detection_rules/config.py:
```diff
@@ -28,7 +28,8 @@ class UnitTest:
     def __post_init__(self):
-        assert not (self.bypass and self.test_only), 'Cannot use both test_only and bypass'
+        assert not (self.bypass is not None and self.test_only is not None), \
+            'Cannot set both `test_only` and `bypass` in test_config!'
```
small nitpick: the double negatives are trickier to understand, so 
```python
assert (self.bypass is None or self.test_only is None)
```
might be better

traut reviewed (approved):
looks good!
